### PR TITLE
Allow Router Middleware for customizing routing

### PR DESCRIPTION
**Pull request by willhug**
Summary: Currently all routing in YARPC is done through a fixed list of procedures
linked to handlers, if a request comes in that is not in the list of procedures it 
is thrown away.  This PR adds the concept of "Router Middleware" that will allow
us to create custom routers for cases where the requests aren't registered on a specific
procedure name.

### CODE REVIEW

File: api/middleware/middlewaretest/router.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Automatically generated by MockGen. DO NOT EDIT!
+// Source: go.uber.org/yarpc/api/middleware (interfaces: Router)
```
Comment (kriskowal) [inline api/middleware/middlewaretest/router.go]: Is there a thing we can do to add `@generated` or whatever it is that Github likes to collapse generated code in diffs? @abhinav

Comment (abhinav) [inline api/middleware/middlewaretest/router.go]: In the past, I have simply added the 'echo'-ed the relevant header to the top
of the file from a script.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,8 @@ Releases
 v1.2.0-dev (unreleased)
 -----------------------
 
+-   Added `RouterMiddleware` parameter to `yarpc.Config`, if provided, this
+    middleware will allow customizing routing to handlers.
```
Comment (abhinav) [inline CHANGELOG.md]: > [..] which, if provided, will allow customizing routing to handlers.

File: api/middleware/router.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"context"
+
+	"go.uber.org/yarpc/api/transport"
+)
+
+// Router is a middleware for defining a customized routing experience for procedures
+type Router interface {
+	// Procedures returns the list of procedures that can be called on this router.
+	// Procedures MUST call into router that is passed in.
```
Comment (abhinav) [inline api/middleware/router.go]: Based on offline discussion, `s/MUST/SHOULD`.

File: api/middleware/router.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	Procedures(transport.Router) []transport.Procedure
+
+	// Choose returns a handlerspec for the given request and transport.
```
Comment (abhinav) [inline api/middleware/router.go]: HandlerSpec

File: api/middleware/router.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	// If the Router cannot determine what to call it should call into the router that was
+	// passed in.
+	Choose(context.Context, *transport.Request, transport.Router) (transport.HandlerSpec, error)
+}
+
+// ApplyRouter applies the given Router middleware to
+// the given Router.
```
Comment (abhinav) [inline api/middleware/router.go]: Inconsistent wrapping on docstrings.

File: api/middleware/router.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+func ApplyRouter(r transport.RouteTable, m Router) transport.RouteTable {
```
Comment (abhinav) [inline api/middleware/router.go]: Consider calling this `ApplyRouteTable` because it operates on a RouteTable, not Router.

**Pull request merged**